src/sgidClient.ts

```diff
@@ -0,0 +1,237 @@
+import { compactDecrypt, importJWK, importPKCS8 } from 'jose'
+import { Client, generators, Issuer } from 'openid-client'
+
+import {
+  API_VERSION,
+  DEFAULT_SCOPE,
+  DEFAULT_SGID_CODE_CHALLENGE_METHOD,
+  SGID_AUTH_METHOD,
+  SGID_SIGNING_ALG,
+  SGID_SUPPORTED_FLOWS,
+} from './constants'
+import * as Errors from './error'
+import {
+  AuthorizationUrlParams,
+  AuthorizationUrlReturn,
+  CallbackParams,
+  SgidClientParams,
+  UserInfoParams,
+} from './types'
+import { convertPkcs1ToPkcs8 } from './util'
+
+export class SgidClient {
+  private privateKey: string
+  private sgID: Client
+
+  /**
+   * Initialises an SgidClient instance.
+   * @param params Constructor arguments
+   * @param params.clientId Client ID provided during client registration
+   * @param params.clientSecret Client secret provided during client registration
+   * @param params.privateKey Client private key provided during client registration
+   * @param params.redirectUri Redirection URI for user to return to your application
+   * after login. If not provided in the constructor, this must be provided to the
+   * authorizationUrl and callback functions.
+   * @param params.hostname Hostname of OpenID provider (sgID). Defaults to
+   * https://api.id.gov.sg.
+   */
+  constructor({
+    clientId,
+    clientSecret,
+    privateKey,
+    redirectUri,
+    hostname = 'https://api.id.gov.sg',
+  }: SgidClientParams) {
+    // TODO: Discover sgID issuer metadata via .well-known endpoint
+    /**
+     * Issuer is appended with version number only from v2 onwards
+     */
+    let issuer = new URL(hostname).origin
+    if ((API_VERSION as number) !== 1) {
+      issuer += `/v${API_VERSION}`
+    }
+
+    const { Client } = new Issuer({
+      issuer,
+      authorization_endpoint: `${hostname}/v${API_VERSION}/oauth/authorize`,
```

Comment:
What's the rationale for using `${hostname}/v${API_VERSION}` instead of just `issuer`

Reply:
Discussed separately: Conclusion is that, while there are some instances where (hostname + version_number) != (issuer), it doesn't apply to us and issuer should always == hostname + version_number. 

Changed to from hostname + version_number to `issuer`